MozillaFoundation/foundation.mozilla.org
# Site-wide Donate banner A/B testing

# Description

The current Donate banner, implemented as a snippet, cannot be A/B tested because our A/B testing package only supports testing at the **Page** level, not for snippets or fields that aren’t directly tied to a page.

This PR enables site-wide A/B testing for the `DonateBanner` by implementing the following changes:

- Added a new Page model, `DonateBannerPage`, which has only 2 fields: `title` and `donate_banner`. This page type is restricted to a single instance in the CMS and can only live under Homepage.

- Updated the `BasePage` model and template. with a new method `get_donate_banner()`:
  - The method retrieves the banner set for the `DonateBannerPage`'s `donate_banner` field.
  - If an active A/B test is running on the `DonateBannerPage`, the method checks if the user has already been assigned a test variant by looking for a specific cookie.
  - If no version cookie is found, a version is returned using the `Abtest` model's `get_new_participant_version()` method, which returns either the "control" or "variant" group.
  - Based on their assigned group, the method returns the appropriate donate banner to display. Additionally, a cookie is set by wagtail-ab-testing to ensure the user continues to see the same version on subsequent visits.

- Introduced a custom Wagtail A/B test event called `DonateBannerLinkClick` which is triggered whenever the banner's CTA/link button is clicked.


# Notes

Here is a list of all pages that either inherit from the `BasePage`, or have been updated to use it's `get_donate_banner()` method. If the donate banner is set on the DonateBannerPage, it will render on the following pages (including during A/B tests):

- Homepage
- AppInstallPage
- ArticlePage
- BanneredCampaignPage
- BlogIndexPage
- BlogPage
- BuyersGuideArticlePage
- BuyersGuideCampaignPage
- BuyersGuideEditorialContentIndexPage
- BuyersGuidePage
- CampaignIndexPage
- CampaignPage
- ConsumerCreepometerPage
- DearInternetPage
- GeneralProductPage
- IndexPage
- InitiativesPage
- MiniSiteNameSpace
- ModularPage
- OpportunityPage
- ParticipatePage2
- PrimaryPage
- ProductPage
- PublicationPage
- RCCAuthorsIndexPage
- RCCDetailPage
- RCCLandingPage
- RCCLibraryPage
- ResearchAuthorsIndexPage
- ResearchDetailPage
- ResearchLandingPage
- ResearchLibraryPage
- Styleguide
- YoutubeRegrets2021Page
- YoutubeRegrets2022Page
- YoutubeRegretsPage
- YoutubeRegretsReporterPage
 

# Screenshots

## DonateBannerPage (CMS):
![Screenshot 2024-11-04 at 21-32-36 Editing Donate banner page Donate Banner Page - Wagtail](https://github.com/user-attachments/assets/42065863-c366-4330-ae6d-cbabb363d72d)

# Steps to test A/B Testing
1. Visit https://foundation-s-donate-ban-93oaup.herokuapp.com/cms/pages/46/
2. Log in with credentials: admin2/admin2
3. Click "Add Child Page" and create a new page of the type "Donate Banner Page"
4. Name the page whatever you would like, and select "CONTROL" for the `donate_banner` field. Publish the page.
5. Confirm that the banner is rendering on https://foundation-s-donate-ban-93oaup.herokuapp.com
6. Edit your donate page again, this time select "VARIANT" for the `donate_banner` field. 
7. Select "save and create AB test"
8. On the A/B test page, select "Donate Banner Link Click" on the "what is the goal?" dropdown. Since the "donate banner link click" event is global, you do not need to choose a page for the green "Choose a page" button. 
9. Click "save and run test"
10. You can now test that a visit to the site will serve you a variant of the donate banner, and that the AB test dashboard is accurately tracking variants served and link clicks 👍
